### DISCUSSION TITLE
Timeline timing question

### DISCUSSION THREAD

**Discussion by Remraf1** [cmu-sei/GHOSTS#99] [Q&A] [answered]
Just wanted to post a couple statements to make sure my understanding of the actions within the timeline file are correct.  Please correct me if I'm wrong.  I often am lol.

1)  In the timeline file, all the handler types are fired off at the same time.  (not counting the delays entered into the TimeLineEvents)
Example, if a timeline had handlers for Word, Chrome, and Command they would all start at execution.  Then the events within the handlers would fire depending on the delays set.

2)  Within each handler type the TimeLineEvents fire off sequentially, separated by the delay timers, and loop that sequence if the Loop variable is True.

Am I close?

Additionally I am having a hard time nailing down the delay variable, doesn't seem to map to seconds or milliseconds.  Also the timer value doesn't seem linear.  A value of 100 is about 1.12 seconds, so I thought a value of 6720 should be about 1 min.  Instead it was about 7.08 seconds.

**Comment by sei-dupdyke** [comment, marked as answer]
You are correct on both (1) and (2) but most handlers (all I think?) have jitter built into them — so that if you have 100 agents doing things on the same timeline, all 100 aren't doing exactly the same thing at the same time.

Delay times are in ms — so 15 minutes is 900000 ms.

I suppose I could make the jitter configurable, so that if you wanted to turn that off, ghosts clients would operate on strict times that you enter on the timeline. But to me, that is a JIT activity, not something that would typically be put into the normal default agent timeline. I might not be thinking about the right use case for what you want to do though, so if its something specific, please let me know.

**Reply by Remraf1** [reply]
Ahh jitter explains what I was seeing.  No adjustments needed, random jitter is perfect thank you!  I was just trying to nail down rough ballpark time spans for my timelines.  Now that I know they are ms, makes this alot easier.  Thank you so much!
